src/sinks/pulsar/service.rs

```diff
@@ -0,0 +1,212 @@
+use lru::LruCache;
+use pulsar::{Error as PulsarError, Executor, Producer, ProducerOptions, Pulsar};
+use tokio::sync::Mutex;
+use tower::Service;
+use vector_common::internal_event::CountByteSize;
+use vector_core::stream::DriverResponse;
+
+use crate::event::{EventFinalizers, EventStatus, Finalizable};
+use crate::internal_events::PulsarSendingError;
+use crate::sinks::pulsar::request_builder::PulsarMetadata;
+use crate::sinks::util::retries::RetryLogic;
+use std::num::NonZeroUsize;
+use vector_common::request_metadata::{MetaDescriptive, RequestMetadata};
+
+#[derive(Clone)]
+pub(super) struct PulsarRequest {
+    pub body: Bytes,
+    pub metadata: PulsarMetadata,
+    pub request_metadata: RequestMetadata,
+}
+
+pub struct PulsarResponse {
+    event_byte_size: usize,
+}
+
+impl DriverResponse for PulsarResponse {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(1, self.event_byte_size)
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.event_byte_size)
+    }
+}
+
+impl Finalizable for PulsarRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.metadata.finalizers)
+    }
+}
+
+impl MetaDescriptive for PulsarRequest {
+    fn get_metadata(&self) -> RequestMetadata {
+        self.request_metadata
+    }
+}
+
+/// Pulsar retry logic.
+#[derive(Debug, Default, Clone)]
+pub struct PulsarRetryLogic;
+
+impl RetryLogic for PulsarRetryLogic {
+    type Error = PulsarError;
+    type Response = PulsarResponse;
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        // TODO improve retry logic
+        true
+    }
+}
+
+type SafeLru<Exe> = Arc<Mutex<LruCache<String, Result<Arc<Mutex<Producer<Exe>>>, PulsarError>>>>;
+
+#[derive(Clone)]
+pub struct PulsarService<Exe: Executor> {
+    pulsar_client: Pulsar<Exe>,
+    producer_cache: SafeLru<Exe>,
+    producer_options: ProducerOptions,
+    producer_name: Option<String>,
+}
+
+impl<Exe: Executor> PulsarService<Exe> {
+    pub(crate) fn new(
+        pulsar_client: Pulsar<Exe>,
+        producer_options: ProducerOptions,
+        producer_cache_size: Option<NonZeroUsize>,
+        producer_name: Option<String>,
+    ) -> PulsarService<Exe> {
+        // Use a LRUCache to store a limited set of producers
+        // Producers in Pulsar use a send buffer, so we want to limit the number of these
+        let producer_cache = Arc::new(Mutex::new(LruCache::new(
+            producer_cache_size.unwrap_or(NonZeroUsize::new(100).unwrap()),
+        )));
+        PulsarService {
+            pulsar_client,
+            producer_cache,
+            producer_options,
+            producer_name,
+        }
+    }
+
+    /// Build a producer that is wrapped in an Arc<Mutex> to allow for the producer
+    /// to control access.
+    ///
+    /// NOTE: Pulsar client library should likely be improved to simplify this
```

Comment:
Worth opening an issue to track this?

Reply:
This logic was added from the original PR author. On a quick scan it's not entirely clear to me why the Arc is necessary. I can look into it more. I checked open/closed issues on the client library and didn't find anything related.

Reply:
It looks to me like the Arc was added because the pulsar crate Producer does not implement Clone.
But I don't think it was necessary to have a cache of producers? because the only thing coming in from the event and being used in the producer creation was the topic... which is set on configuration. I suspect that may have been a misunderstanding from the original author, unless I overlooked something entirely.

Reply:
I isolated this change (mostly) in this commit so it is easier to see what was done: https://github.com/vectordotdev/vector/pull/14345/commits/19f948700d83ac2966a86a521c01ff076af9c353

Reply:
🤔 I vaguely remember this, I'll take a look at the commit and try and remember more.

Reply:
just closing the loop that we believe to have resolved the need for a manual cache by using a multi producer struct that the pulsar crate provides.